# Microarrays {#microarrays}

```{r global-options, include=FALSE}
library(knitr)
opts_chunk$set(results='hide', warning=FALSE, message=FALSE)
```

```{r libraries}
# Publication-related
library(kableExtra)

# Bioconductor
library(GEOquery)
library(oligo)
library(limma)
library(biobroom)

# Domain specific
library(rentrez)

# General purpose
library(curl)
library(fs)
library(urltools)
library(xml2)

# Tidyverse
library(lubridate)
library(janitor)
library(tidyverse)
```

## Overview

Our goal is to find some measurement we can perform to distinguish T cells from other cells, as well as T cell subsets from one another. Gene expression as measured by expression microarrays is one candidate measurement we could perform. In this chapter, we'll seek out open expression microarray data generated from T cells and do some analysis to see if we can distinguish T cells and their subsets using this data.

We will create several tibbles over the course of this chapter, including:

1. `c7_tcells`: our best attempt at pulling out the T cell data sets from ImmuneSigDB. Each row is a gene set.

1. `c7_pubs`: publications from which ImmuneSigDB gene sets are derived. Each row is a PMID.

1. `pGSE2770tidy`: phenotype information for a single GSE GEO ID from ImmuneSigDB. Each row is a GSM GEO ID, i.e. a sample.

1. `eset_tib_genes`: preprocessed expression data for the samples we're analyzing. "Wide" version of table with one column per sample. Primary key is gene.

1. `eset_tidy_p`: preprocessed expression data joined to phenotype data about the sample. "Tall" version of table. Primary key is (gene, sample).

1. `results_genes_tib`: results of the limma differential expression analysis. Each row is a gene.

## Expression profiling with microarrays

Explain how microarrays work. Probably reference opening of RMA paper.

## T cell expression profiles

My first thought on where to look for expression data on T cells: ImmuneSigDB.

## Exploring ImmuneSigDB

What's in ImmuneSigDB? To find out, head over to the [GSEA downloads page](http://software.broadinstitute.org/gsea/downloads.jsp) (you'll need to register your email address first) and grab the "Current MSigDB xml file"; at the time of this writing, that's `msigdb_v6.2.xml`.

Let's put all of our downloaded data into a single directory.

```{r create-data-dir}
data_dir <- dir_create("data")
```

Put the MSigDB file into this directory at `msigdb_files/msigdb_v6.2.xml`. We are only interested in the C7 gene set, also known as ImmuneSigDB. For now we are using XPath to extract the relevant data; perhaps we should show how to explore the XML file and construct this query.

```{r extract-c7, dependson='create-data-dir'}
x <- read_xml(fs::path(data_dir, "msigdb_files/msigdb_v6.2.xml"))
c7_xml <- xml_find_all(x, "//GENESET[contains(@CATEGORY_CODE,'C7')]")
c7_tib <- c7_xml %>% as_list %>% purrr::map(~ as_tibble(attributes(.x))) %>% bind_rows %>% clean_names
```

Each row in `c7_tib` represents a gene set from ImmuneSigDB. We filter down to those gene sets we suspect relate to T cells.

```{r filter-tcells, dependson='extract-c7'}
tcell_terms <- c("TCELL", "CD3", "CD4", "CD8")
c7_tcells <- c7_tib %>% rowwise %>% filter(any(str_detect(standard_name, tcell_terms))) %>% ungroup
```

We can now explore publications and GEO data sets used to build these gene sets. We are specifically interested in human, not mouse, T cells in this chapter. 

```{r tcell-pubs, dependson='filter-tcells'}
tcell_pubs <- c7_tcells %>%
  filter(organism == "Homo sapiens") %>%
  group_by(pmid) %>%
  summarize(n_genesets = n(), n_geoids = n_distinct(geoid), geoids = toString(unique(geoid))) %>%
  mutate(n_geoids = if_else(geoids == "", 0L, n_geoids)) %>%
  arrange(desc(n_genesets))
```

Let's get some metdata for these papers. 

```{r pub-metadata, dependson='tcell-pubs'}
pub_meta_esum <- entrez_summary(db = "pubmed", id = tcell_pubs$pmid, config = httr::config(http_version = 2))
pub_meta_list <- extract_from_esummary(pub_meta_esum, c("uid", "pubdate", "title")) %>% array_branch(2)
pub_meta_vec <- pub_meta_list %>% purrr::map(unlist) %>% flatten_chr()
pub_meta <- tribble(~pmid, ~pubdate, ~title, !!!pub_meta_vec)
pub_meta <- pub_meta %>% mutate(pubdate = parse_date_time(pubdate, orders = c("Y", "Ym", "Ymd")))
tcell_pubs <- tcell_pubs %>% left_join(pub_meta, by = "pmid")
```

```{r tcell-pubs-table, dependson='pub-metadata', fig.cap = "T cell publications", echo = FALSE, results = "markup"}
tcell_pubs %>%
  mutate(
    pmid = text_spec(pmid, link = paste0("https://www.ncbi.nlm.nih.gov/pubmed/", pmid))
  ) %>%
  kable(escape = FALSE) %>%
  column_spec(5, width = "10em") %>%
  column_spec(6, width = "1000em")
```



## Data from one paper
First, we download data for GSE2770 from [GEO](https://www.ncbi.nlm.nih.gov/geo/). There are 3 elements in this list, 1 for each platform.

```{r getgeo-gse2770, dependson='create-data-dir'}
download_dir <- fs::path(data_dir, "series_matrix_files")
if (!dir_exists(download_dir)) {
  dir_create(download_dir)
  GSE2770 <- getGEO("GSE2770", getGPL=FALSE, destdir=download_dir)
} else {
  GSE2770 <- dir_ls(download_dir, glob="*series_matrix*") %>%
    purrr::map(~ getGEO(filename=.x, getGPL=FALSE))
}
```

Next, we pull out the phenotype data.

```{r pgse2770, dependson='getgeo-gse2770'}
pGSE2770 <- GSE2770 %>%
  purrr::map(~ as(pData(.x), "data.frame")) %>%
  bind_rows() %>%
  as_tibble()
```

Then we tidy up the phenotype data.

```{r pgse2770tidy, dependson='pgse2770'}
pGSE2770tidy <- pGSE2770 %>%
  select(platform_id, geo_accession, supplementary_file, title) %>%
  separate(title, c("cells", "treatment", "time", "replicate", "platform"), sep="[_\\(\\)]") %>%
  select(-cells) %>%
  mutate(replicate = map_chr(replicate, ~ str_split(.x, boundary("word"))[[1]][2])) %>%
  mutate(time = as.duration(time))
```

## Data from one gene set

Let's just get the CEL files used to generate the [GSE2770_IL12_AND_TGFB_ACT_VS_ACT_CD4_TCELL_6H_DN](http://software.broadinstitute.org/gsea/msigdb/geneset_page.jsp?geneSetName=GSE2770_IL12_AND_TGFB_ACT_VS_ACT_CD4_TCELL_6H_DN) gene set from ImmuneSigDB [@Godec2016-ku]. There are only two treatments to compare, at a single time point. Each treatment is replicated twice.

```{r get-cel-list, dependson='pgse2770tidy'}
# get just the U133A cel files
cel_file_urls <- pGSE2770tidy %>%
  filter(time == duration("6h") & (treatment %in% c("antiCD3+antiCD28+IL12+TGFbeta", "antiCD3+antiCD28")) & platform == "U133A") %>%
  .$supplementary_file
```

We then download the CEL files.

```{r download-cels, dependson='get-cel-list'}
download_dir <- fs::path(data_dir, "cel_files")
if (!dir_exists(download_dir)) {
  dir_create(download_dir)
  cel_file_names <- path_file(url_parse(cel_file_urls)$path)
  cel_files_local <- map2(cel_file_urls, fs::path(download_dir, cel_file_names), ~ curl_download(.x, .y)) %>% simplify()
} else {
  cel_files_local <- dir_ls(download_dir, glob="*.cel*", ignore.case=TRUE)
}
```

## Load expression data

And read them into an oligo ExpressionFeatureSet object. Note the feature data (i.e. the probeset IDs) are not stored in the feature data of the ExpressionSet object, but rather in a SQLlite database [pd.hg.u133a](https://bioconductor.org/packages/release/data/annotation/html/pd.hg.u133a.html) somewhere on disk that the call to `rma()` will pick up.

```{r read-cels, dependson='download-cels'}
raw_data <- read.celfiles(cel_files_local)
```

## Preprocess expression data

Run RMA on our batch of CEL files to be compared.

```{r rma, dependson='read-cels'}
processed_data <- rma(raw_data)
as_tibble(exprs(processed_data), rownames = "probe_set_id")
```

## Map probesets to genes

Because we want to do our differential expression analysis at the gene level, not the probeset level, we need to map probeset IDs to gene symbols. To ensure comparability with the GSEA gene sets, we'll use the annotations provided by the Broad that are used in GSEA.

```{r download-chip}
chip_file_url <- "ftp://ftp.broadinstitute.org/pub/gsea/annotations/HG_U133A.chip"
chip_file_name <- path_file(url_parse(chip_file_url)$path)

download_dir <- fs::path(data_dir, "chip_files")
if (!dir_exists(download_dir)) {
  dir_create(download_dir)
  chip_file <- curl_download(chip_file_url, fs::path(download_dir, chip_file_name))
} else {
  chip_file <- fs::path(download_dir, chip_file_name)
}
```

We need to clean up the data a bit: the import produced a bogus column (X4), and we only want one gene symbol per probeset ID.

```{r tidy-chip, dependson='download-chip'}
U133A_chip_raw <- read_tsv(chip_file)
U133A_chip <- U133A_chip_raw %>% 
  select(-X4) %>%
  clean_names() %>%
  rowwise() %>%
  mutate(first_gene_symbol = str_trim(str_split(gene_symbol, "///")[[1]][1])) 
```

Finally we can join with our processed data!

```{r join-chip, dependson=c('rma', 'download-chip')}
eset_tib <- as_tibble(exprs(processed_data), rownames = "probe_set_id")
eset_tib_genes <- eset_tib %>% left_join(U133A_chip, by = "probe_set_id")
```

We now remove probesets that don't map to a gene symbol; note chip files encode this fact with `---`, which we demonstrate first before the filter.

```{r remove-unmapped-probesets, dependson='join-chip'}
eset_tib_genes %>% filter(is.na(first_gene_symbol)) %>% count() # 0
eset_tib_genes %>% filter(first_gene_symbol == "---") %>% count() # 1,109
eset_tib_genes <- eset_tib_genes %>% filter(first_gene_symbol != "---")
```

Some genes have multiple probesets that map to them. GSEA handles this situation by only keeping the maximum intensity value across all probesets that map to the gene. In GSEA terminology, we are using the max probe algorithm to collapse our pobesets at the gene level.

```{r collapse-probesets, dependson='remove-unmapped-probesets'}
eset_tib_genes_only <- eset_tib_genes %>%
  select(-c(probe_set_id, gene_symbol, gene_title)) %>%
  group_by(first_gene_symbol) %>%
  summarise_all(max)
```

## Exploring preprocessed expression data

We join our expression data with our phenotype data so that we know which treatment was used for each sample.

```{r explore-eset, dependson='collapse-probesets'}
eset_tidy <- eset_tib_genes_only %>%
  gather(key = "geo_accession", value = "expression", starts_with("GSM")) %>%
  mutate(geo_accession = str_extract(geo_accession, "^([^.]+)"))
eset_tidy_p <- eset_tidy %>%
  left_join(pGSE2770tidy, by = "geo_accession")
```

```{r graph-genes, dependson='explore-eset', fig.cap = "Expression level vs. treatment"}
genes <- c("IFNG", "TBX21", "IL4", "GATA3")
ggplot(eset_tidy_p %>% filter(first_gene_symbol %in% genes),
       aes(x=treatment, y=expression, color=replicate)) +
  geom_point() +
  facet_wrap(~first_gene_symbol, ncol=1, scales="free") +
  theme_bw()
```


## Differential expression analysis

Now we'll take our new genes-only expression data and put it into a form expected by `limma`.

```{r make-expressionset, dependson='collapse-probesets'}
eset_genes <- eset_tib_genes_only %>%
  as.data.frame() %>%
  column_to_rownames(var="first_gene_symbol") %>%
  data.matrix() %>%
  ExpressionSet()
```

One last step: we need to make the design matrix for this comparison.

```{r make-design}
treatments <- factor(c(1,2,2,1), labels=c("untreated", "treated")) # weird ordering of files
design <- model.matrix(~treatments)
```

Finally we use `limma` to perform a differential expression analysis!

```{r do-dge, dependson=c('make-expressionset', 'make-design')}
fit_genes <- lmFit(eset_genes, design)
fit2_genes <- eBayes(fit_genes)
results_genes <- topTable(fit2_genes, number=Inf)
results_genes_tib <- as_tibble(results_genes, rownames = "gene")
```

Let's look at our results.

```{r volcano, dependson='do-dge', fig.cap = "Volcano plot made with biobroom", echo = FALSE}
ggplot(tidy(fit2_genes), aes(x=estimate, y=log(p.value), color=statistic)) + 
  geom_point() +
  theme_bw()
```