<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Learning About T Cells Through Data</title>
  <meta name="description" content="An examination of T cells using several open data sets.">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Learning About T Cells Through Data" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="images/cover.png" />
  <meta property="og:description" content="An examination of T cells using several open data sets." />
  <meta name="github-repo" content="hammerlab/learning-about-t-cells-through-data" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Learning About T Cells Through Data" />
  
  <meta name="twitter:description" content="An examination of T cells using several open data sets." />
  <meta name="twitter:image" content="images/cover.png" />

<meta name="author" content="Hammer Lab">


<meta name="date" content="2018-10-12">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction.html">
<link rel="next" href="microarrays.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/kePrint-0.0.1/kePrint.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="flow.html"><a href="flow.html"><i class="fa fa-check"></i><b>2</b> Flow</a><ul>
<li class="chapter" data-level="2.1" data-path="flow.html"><a href="flow.html#exploring-flowrepository"><i class="fa fa-check"></i><b>2.1</b> Exploring FlowRepository</a></li>
<li class="chapter" data-level="2.2" data-path="flow.html"><a href="flow.html#data-from-one-paper"><i class="fa fa-check"></i><b>2.2</b> Data from one paper</a><ul>
<li class="chapter" data-level="2.2.1" data-path="flow.html"><a href="flow.html#gating"><i class="fa fa-check"></i><b>2.2.1</b> Gating</a></li>
<li class="chapter" data-level="2.2.2" data-path="flow.html"><a href="flow.html#results"><i class="fa fa-check"></i><b>2.2.2</b> Results</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="microarrays.html"><a href="microarrays.html"><i class="fa fa-check"></i><b>3</b> Microarrays</a><ul>
<li class="chapter" data-level="3.1" data-path="microarrays.html"><a href="microarrays.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="microarrays.html"><a href="microarrays.html#expression-profiling-with-microarrays"><i class="fa fa-check"></i><b>3.2</b> Expression profiling with microarrays</a></li>
<li class="chapter" data-level="3.3" data-path="microarrays.html"><a href="microarrays.html#t-cell-expression-profiles"><i class="fa fa-check"></i><b>3.3</b> T cell expression profiles</a></li>
<li class="chapter" data-level="3.4" data-path="microarrays.html"><a href="microarrays.html#exploring-immunesigdb"><i class="fa fa-check"></i><b>3.4</b> Exploring ImmuneSigDB</a></li>
<li class="chapter" data-level="3.5" data-path="microarrays.html"><a href="microarrays.html#data-from-one-paper-1"><i class="fa fa-check"></i><b>3.5</b> Data from one paper</a></li>
<li class="chapter" data-level="3.6" data-path="microarrays.html"><a href="microarrays.html#data-from-one-gene-set"><i class="fa fa-check"></i><b>3.6</b> Data from one gene set</a></li>
<li class="chapter" data-level="3.7" data-path="microarrays.html"><a href="microarrays.html#load-expression-data"><i class="fa fa-check"></i><b>3.7</b> Load expression data</a></li>
<li class="chapter" data-level="3.8" data-path="microarrays.html"><a href="microarrays.html#preprocess-expression-data"><i class="fa fa-check"></i><b>3.8</b> Preprocess expression data</a></li>
<li class="chapter" data-level="3.9" data-path="microarrays.html"><a href="microarrays.html#map-probesets-to-genes"><i class="fa fa-check"></i><b>3.9</b> Map probesets to genes</a></li>
<li class="chapter" data-level="3.10" data-path="microarrays.html"><a href="microarrays.html#exploring-preprocessed-expression-data"><i class="fa fa-check"></i><b>3.10</b> Exploring preprocessed expression data</a></li>
<li class="chapter" data-level="3.11" data-path="microarrays.html"><a href="microarrays.html#differential-expression-analysis"><i class="fa fa-check"></i><b>3.11</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="rnaseq.html"><a href="rnaseq.html"><i class="fa fa-check"></i><b>4</b> RNA-seq</a></li>
<li class="chapter" data-level="5" data-path="scrnaseq.html"><a href="scrnaseq.html"><i class="fa fa-check"></i><b>5</b> scRNA-seq</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Learning About T Cells Through Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="flow" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Flow</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Bioconductor</span>
<span class="kw">library</span>(FlowRepositoryR)
<span class="kw">library</span>(flowCore)
<span class="kw">library</span>(openCyto)
<span class="kw">library</span>(ggcyto)

<span class="co"># Tidyverse</span>
<span class="kw">library</span>(fs)
<span class="kw">library</span>(tidyverse)</code></pre></div>
<p>We’ll examine the OMIP panel data available from <a href="https://flowrepository.org/">FlowRepository</a>, using the client library <a href="https://bioconductor.org/packages/release/bioc/html/FlowRepositoryR.html">FlowRepositoryR</a>, and demonstrate phenotypic differences between T cell types based on the protein expression levels captured in these assays.</p>
<p>To aid in identifying and visualizing these different cell populations, a variety of tools built on <a href="https://bioconductor.org/packages/release/bioc/html/FlowRepositoryR.html">flowCore</a> will be used and in particular, <a href="https://bioconductor.org/packages/release/bioc/html/openCyto.html">openCyto</a> will enable us to model and manipulate the workflows used in the OMIP publications to identify T cell types.</p>
<div id="exploring-flowrepository" class="section level2">
<h2><span class="header-section-number">2.1</span> Exploring FlowRepository</h2>
<p>TBD</p>
<ul>
<li>Attempt to show consolidated metadata from all OMIP datasets</li>
<li>This will likely need to involve downloading the data to demonstrate anything useful (i.e. which markers are present)</li>
<li>Show markers available by datasets relating to Human t-cells (not sure how to filter to human only yet)</li>
</ul>
</div>
<div id="data-from-one-paper" class="section level2">
<h2><span class="header-section-number">2.2</span> Data from one paper</h2>
<p>Initial Example: <a href="https://flowrepository.org/id/FR-FCM-ZZ9H">OMIP-021 Innate-like T-cell Panel (FR-FCM-ZZ9H)</a></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_dir &lt;-<span class="st"> </span><span class="kw">dir_create</span>(<span class="st">&quot;data/flow&quot;</span>)
dataset_id &lt;-<span class="st"> &quot;FR-FCM-ZZ9H&quot;</span>
dataset_path &lt;-<span class="st"> </span>fs<span class="op">::</span><span class="kw">path</span>(data_dir, dataset_id)
dataset &lt;-<span class="st"> </span><span class="kw">flowRep.get</span>(dataset_id)

<span class="cf">if</span> (<span class="op">!</span><span class="kw">dir_exists</span>(dataset_path))
  dataset &lt;-<span class="st"> </span><span class="kw">download</span>(dataset, <span class="dt">dirpath=</span>dataset_path, <span class="dt">only.files=</span><span class="st">&quot;Donor.*fcs&quot;</span>, <span class="dt">show.progress=</span>F)
fr &lt;-<span class="st"> </span><span class="kw">read.FCS</span>(fs<span class="op">::</span><span class="kw">path</span>(<span class="kw">getwd</span>(), dataset_path, <span class="st">&#39;Donor1.fcs&#39;</span>))
fr <span class="op">%&gt;%</span><span class="st"> </span>exprs <span class="op">%&gt;%</span><span class="st"> </span>data.frame <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>(<span class="dv">10</span>) <span class="op">%&gt;%</span><span class="st"> </span>knitr<span class="op">::</span><span class="kw">kable</span>(<span class="dt">format.args =</span> <span class="kw">list</span>(<span class="dt">digits=</span><span class="dv">3</span>))</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">FSC.A</th>
<th align="right">FSC.H</th>
<th align="right">SSC.A</th>
<th align="right">SSC.H</th>
<th align="right">X450.50Violet.A</th>
<th align="right">X525.50Violet.A</th>
<th align="right">X540.30Violet.A</th>
<th align="right">X585.15Violet.A</th>
<th align="right">X610.20Violet.A</th>
<th align="right">X670.30Violet.A</th>
<th align="right">X670.14Red.A</th>
<th align="right">X730..45Red.A</th>
<th align="right">X780.60Red.A</th>
<th align="right">X530.30Blue.A</th>
<th align="right">X710.50Blue.A</th>
<th align="right">X582.15Yellow.A</th>
<th align="right">X610.20Yellow.A</th>
<th align="right">X670.30Yellow.A</th>
<th align="right">X710.50Yellow.A</th>
<th align="right">X780.60Yellow.A</th>
<th align="right">Time</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">119565</td>
<td align="right">108322</td>
<td align="right">82937</td>
<td align="right">76142</td>
<td align="right">1130</td>
<td align="right">257</td>
<td align="right">630</td>
<td align="right">1002</td>
<td align="right">3625</td>
<td align="right">5465</td>
<td align="right">7061</td>
<td align="right">13215</td>
<td align="right">6966</td>
<td align="right">222.4</td>
<td align="right">17786</td>
<td align="right">803</td>
<td align="right">418</td>
<td align="right">20700</td>
<td align="right">29263</td>
<td align="right">19368</td>
<td align="right">136</td>
</tr>
<tr class="even">
<td align="right">107004</td>
<td align="right">77596</td>
<td align="right">262143</td>
<td align="right">207144</td>
<td align="right">1223</td>
<td align="right">732</td>
<td align="right">995</td>
<td align="right">169</td>
<td align="right">721</td>
<td align="right">114</td>
<td align="right">1110</td>
<td align="right">8055</td>
<td align="right">2695</td>
<td align="right">420.8</td>
<td align="right">926</td>
<td align="right">358</td>
<td align="right">294</td>
<td align="right">851</td>
<td align="right">70563</td>
<td align="right">2586</td>
<td align="right">136</td>
</tr>
<tr class="odd">
<td align="right">214908</td>
<td align="right">139281</td>
<td align="right">262143</td>
<td align="right">226353</td>
<td align="right">3218</td>
<td align="right">1049</td>
<td align="right">2240</td>
<td align="right">1467</td>
<td align="right">7903</td>
<td align="right">10092</td>
<td align="right">8914</td>
<td align="right">26039</td>
<td align="right">10930</td>
<td align="right">561.1</td>
<td align="right">19171</td>
<td align="right">994</td>
<td align="right">733</td>
<td align="right">21668</td>
<td align="right">88813</td>
<td align="right">12031</td>
<td align="right">136</td>
</tr>
<tr class="even">
<td align="right">119453</td>
<td align="right">107496</td>
<td align="right">67842</td>
<td align="right">62072</td>
<td align="right">2473</td>
<td align="right">351</td>
<td align="right">5828</td>
<td align="right">14664</td>
<td align="right">22003</td>
<td align="right">3548</td>
<td align="right">7316</td>
<td align="right">6115</td>
<td align="right">951</td>
<td align="right">112.0</td>
<td align="right">5504</td>
<td align="right">184</td>
<td align="right">448</td>
<td align="right">4450</td>
<td align="right">17505</td>
<td align="right">1143</td>
<td align="right">136</td>
</tr>
<tr class="odd">
<td align="right">122345</td>
<td align="right">111352</td>
<td align="right">62728</td>
<td align="right">56540</td>
<td align="right">541</td>
<td align="right">194</td>
<td align="right">2601</td>
<td align="right">735</td>
<td align="right">1267</td>
<td align="right">1095</td>
<td align="right">3060</td>
<td align="right">4466</td>
<td align="right">4531</td>
<td align="right">95.4</td>
<td align="right">5793</td>
<td align="right">435</td>
<td align="right">440</td>
<td align="right">8354</td>
<td align="right">16969</td>
<td align="right">3564</td>
<td align="right">137</td>
</tr>
<tr class="even">
<td align="right">137030</td>
<td align="right">124521</td>
<td align="right">77133</td>
<td align="right">70557</td>
<td align="right">4062</td>
<td align="right">471</td>
<td align="right">10052</td>
<td align="right">23228</td>
<td align="right">41648</td>
<td align="right">6317</td>
<td align="right">13904</td>
<td align="right">15326</td>
<td align="right">2217</td>
<td align="right">152.7</td>
<td align="right">5801</td>
<td align="right">189</td>
<td align="right">829</td>
<td align="right">10039</td>
<td align="right">21626</td>
<td align="right">2003</td>
<td align="right">138</td>
</tr>
<tr class="odd">
<td align="right">153801</td>
<td align="right">122139</td>
<td align="right">122174</td>
<td align="right">80251</td>
<td align="right">924</td>
<td align="right">735</td>
<td align="right">2110</td>
<td align="right">1059</td>
<td align="right">2827</td>
<td align="right">2641</td>
<td align="right">287</td>
<td align="right">1277</td>
<td align="right">611</td>
<td align="right">4595.7</td>
<td align="right">6182</td>
<td align="right">175</td>
<td align="right">201</td>
<td align="right">822</td>
<td align="right">33043</td>
<td align="right">1301</td>
<td align="right">138</td>
</tr>
<tr class="even">
<td align="right">177071</td>
<td align="right">145150</td>
<td align="right">168704</td>
<td align="right">148426</td>
<td align="right">731</td>
<td align="right">679</td>
<td align="right">1243</td>
<td align="right">337</td>
<td align="right">1045</td>
<td align="right">201</td>
<td align="right">415</td>
<td align="right">9995</td>
<td align="right">1659</td>
<td align="right">291.3</td>
<td align="right">1052</td>
<td align="right">271</td>
<td align="right">366</td>
<td align="right">570</td>
<td align="right">47515</td>
<td align="right">953</td>
<td align="right">138</td>
</tr>
<tr class="odd">
<td align="right">165853</td>
<td align="right">118245</td>
<td align="right">122113</td>
<td align="right">89253</td>
<td align="right">2785</td>
<td align="right">511</td>
<td align="right">7996</td>
<td align="right">2560</td>
<td align="right">15636</td>
<td align="right">9627</td>
<td align="right">10654</td>
<td align="right">29089</td>
<td align="right">14198</td>
<td align="right">337.8</td>
<td align="right">20733</td>
<td align="right">1203</td>
<td align="right">2341</td>
<td align="right">24269</td>
<td align="right">40187</td>
<td align="right">12859</td>
<td align="right">138</td>
</tr>
<tr class="even">
<td align="right">121005</td>
<td align="right">108937</td>
<td align="right">72537</td>
<td align="right">66157</td>
<td align="right">255</td>
<td align="right">205</td>
<td align="right">3215</td>
<td align="right">724</td>
<td align="right">2316</td>
<td align="right">153</td>
<td align="right">361</td>
<td align="right">829</td>
<td align="right">141</td>
<td align="right">166.0</td>
<td align="right">4320</td>
<td align="right">335</td>
<td align="right">819</td>
<td align="right">1027</td>
<td align="right">21945</td>
<td align="right">244</td>
<td align="right">138</td>
</tr>
</tbody>
</table>
<p>In this dataset, one way to identify lymphocytes is by looking at modes in the relationship between side and forward scatter:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggcyto</span>(fr, <span class="kw">aes</span>(<span class="dt">x=</span><span class="st">&#39;FSC-A&#39;</span>, <span class="dt">y=</span><span class="st">&#39;SSC-A&#39;</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_hex</span>(<span class="dt">bins=</span><span class="dv">100</span>)</code></pre></div>
<p><img src="Flow_files/figure-html/example-biplot-1.png" width="672" /></p>
<p>To identify the lymphocyte cells above, <code>openCyto</code> gating can be used to select the largest cluster of cells automatically and visualize what fraction of the population these cells constitute:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gate &lt;-<span class="st"> </span>openCyto<span class="op">::</span><span class="kw">flowClust.2d</span>(fr, <span class="st">&#39;FSC-A&#39;</span>, <span class="st">&#39;SSC-A&#39;</span>, <span class="dt">K=</span><span class="dv">3</span>, <span class="dt">quantile=</span>.<span class="dv">9</span>)
<span class="kw">ggcyto</span>(fr, <span class="kw">aes</span>(<span class="dt">x=</span><span class="st">&#39;FSC-A&#39;</span>, <span class="dt">y=</span><span class="st">&#39;SSC-A&#39;</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_hex</span>(<span class="dt">bins=</span><span class="dv">100</span>) <span class="op">+</span><span class="st"> </span><span class="kw">geom_gate</span>(gate) <span class="op">+</span><span class="st"> </span><span class="kw">geom_stats</span>()</code></pre></div>
<p><img src="Flow_files/figure-html/plot-fsc-ssc-1.png" width="672" /></p>
<div id="gating" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Gating</h3>
<p>By repeating the above process for all cell types of interest in the paper, the workflow can be reproduced via a <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/openCyto/inst/doc/HowToWriteCSVTemplate.html">GatingTemplate</a> as shown below, which could also be built programmatically instead. The graphical representation of the template shows how cell types will be recursively defined based on 2 dimensional filters:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">template &lt;-<span class="st"> &#39;alias,pop,parent,dims,gating_method,gating_args,collapseDataForGating,groupBy,preprocessing_method,preprocessing_args</span>
<span class="st">Live,+,root,&quot;FSC-A,525/50Violet-A&quot;,&quot;polyGate&quot;,&quot;x=c(0,3e5,3e5,1e5,.5e5,0),y=c(0,0,2.3,2.3,2,1.5)&quot;,,,,</span>
<span class="st">Lymphocytes,+,Live,&quot;FSC-A,SSC-A&quot;,&quot;flowClust.2d&quot;,&quot;K=3,quantile=.95&quot;,,,,</span>
<span class="st">Singlets,+,Lymphocytes,&quot;FSC-A,FSC-H&quot;,&quot;singletGate&quot;,&quot;maxit=1000,wider_gate=T,prediction_level=.999999999&quot;,,,,</span>
<span class="st">gdTCells,+,Singlets,&quot;670/30Violet-A,530/30Blue.A&quot;,&quot;flowClust.2d&quot;,&quot;K=3,target=c(2.5,2.5)&quot;,,,,</span>
<span class="st">abTCells,+,Singlets,&quot;670/30Violet-A,530/30Blue.A&quot;,&quot;flowClust.2d&quot;,&quot;K=3,target=c(2.5,1),quantile=0.95&quot;,,,,</span>
<span class="st">maiTCells,+,abTCells,&quot;582/15Yellow.A,540/30Violet.A&quot;,&quot;polyGate&quot;,&quot;x=c(2.7,5,5,2.7),y=c(2.5,2.5,5,5)&quot;,,,,&#39;</span>
template_path &lt;-<span class="st"> </span><span class="kw">file_temp</span>(<span class="dt">ext=</span><span class="st">&#39;.csv&#39;</span>)
<span class="kw">write_lines</span>(template, template_path)
gt &lt;-<span class="st"> </span><span class="kw">gatingTemplate</span>(template_path)
<span class="kw">plot</span>(gt)</code></pre></div>
<p><img src="Flow_files/figure-html/plot-gating-template-1.png" width="672" /></p>
<p>The above template only outlines the gating workflow but to apply it to our data, these are the common steps:</p>
<ul>
<li>Apply compensation to the raw FCS data if necessary (not necessary in this case as the authors did this beforehand)</li>
<li>Define channel transformations to make gating and visualization possible</li>
<li>Define any custom gating functions needed in the workflow, which is particularly useful for setting manual gates</li>
<li>Apply the gating template to the data at hand, which in this case is represented as a <code>flowFrame</code> but could also be a <code>flowSet</code> representing a collection of experiments</li>
</ul>
<p>Here is the realization of these steps for this data specifically:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Define logical transformation for all fluorescent channels and build a &quot;GatingSet&quot;, which is a wrapper</span>
<span class="co"># class that binds numeric data with transformations and gating information</span>
transformer &lt;-<span class="st"> </span><span class="kw">transformerList</span>(<span class="kw">colnames</span>(fr<span class="op">@</span>description<span class="op">$</span>SPILL), <span class="kw">logicle_trans</span>())
gs &lt;-<span class="st"> </span><span class="kw">transform</span>(<span class="kw">GatingSet</span>(<span class="kw">flowSet</span>(fr)), transformer)

<span class="co"># Define a custom polygon gating function, that is used in our template to deal with situations</span>
<span class="co"># that are difficult to define an automated gate for</span>
.polyGate &lt;-<span class="st"> </span><span class="cf">function</span>(fr, pp_res, channels, <span class="dt">filterId=</span><span class="st">&quot;polygate&quot;</span>, ...){
  args &lt;-<span class="st"> </span><span class="kw">list</span>(...)
  g &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">x=</span>args<span class="op">$</span>x, <span class="dt">y=</span>args<span class="op">$</span>y)
  <span class="kw">colnames</span>(g) &lt;-<span class="st"> </span>channels
  flowCore<span class="op">::</span><span class="kw">polygonGate</span>(<span class="dt">.gate=</span>g, <span class="dt">filterId=</span>filterId)
}
<span class="kw">registerPlugins</span>(<span class="dt">fun=</span>.polyGate, <span class="dt">methodName=</span><span class="st">&#39;polyGate&#39;</span>,<span class="dt">dep=</span><span class="ot">NA</span>)

<span class="co"># Apply the gating to the data (this may take a couple minutes)</span>
<span class="kw">gating</span>(gt, gs)</code></pre></div>
</div>
<div id="results" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Results</h3>
<p>Now that the workflow is finished, here is a look at all cell types identified:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">autoplot</span>(gs[[<span class="dv">1</span>]], <span class="dt">strip.text =</span> <span class="st">&quot;gate&quot;</span>, <span class="dt">bins=</span><span class="dv">100</span>, <span class="dt">merge=</span>F, <span class="dt">axis_inverse_trans=</span>F) <span class="op">+</span><span class="st"> </span>flow_theme</code></pre></div>
<p><img src="Flow_files/figure-html/plot-cell-types-1.png" width="768" /></p>
<p>This should then be comparable to what was in the <a href="https://onlinelibrary.wiley.com/doi/full/10.1002/cyto.a.22475">OMIP-021 publication</a>, and here are relevant figures demonstrating the similarity of the cell subsets captured:</p>
<div class="figure">
<img src="https://drive.google.com/uc?export=view&amp;id=1LJ9kSIHO-PnqOy3BnP1cy0t93uw2BMrF" alt="optional caption text" />
<p class="caption">optional caption text</p>
</div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="microarrays.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": false,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
